# Costa Rican Household Poverty Level Prediction

Costa Rican Household Poverty Level Prediction is a Kaggle competition which is currently open for late submissions. This competition is for a social cause to help Inter-American Development Bank to identify which households have the highest need for social welfare assistance. Many social programs have a hard time making sure the right people are given enough aid. It’s especially tricky when a program focuses on the poorest segment of the population. The world’s poorest typically can’t provide the necessary income and expense records to prove that they qualify. Other than Costa Rica, many other countries also face this same problem of inaccurately assessing social need. If data scientists can generate an improvement, many countries can take benefit out of it.

The dataset available in Kaggle is a set of household characteristics from a representative sample of Costa Rican Households. The dataset has observations for each member of the household but the classification is done at the household level. This is a multi-class classification problem using a supervised machine learning approach. We classify households according to their income levels (1 = extreme poverty,  2 = moderate poverty, 3 = vulnerable households, 4 = non-vulnerable households). We have identified the multiclass classification algorithms which we think apt to use this type of problems like, KNN, Logistic Regression, Naive Bayes, Random Forest etc. We will fine-tune the hyperparameters for each model and evaluate the different models using f1-score and accuracy metrics to find out the top best among them. Will do significance tests to figure out the optimal model to perform prediction.

Kaggle competition link:

https://www.kaggle.com/c/costa-rican-household-poverty-prediction (Links to an external site.)Links to an external site.
  

# Setup

First, let's make sure this notebook works well in both python 2 and 3, import a few common modules, ensure MatplotLib plots figures inline.

In [3]:
# To support both python 2 and python 3
from __future__ import division, print_function, unicode_literals

# Common imports
import numpy as np
import os

# to make this notebook's output stable across runs
np.random.seed(42)

# To plot pretty figures
%matplotlib inline
import matplotlib
import matplotlib.pyplot as plt
plt.rcParams['axes.labelsize'] = 14
plt.rcParams['xtick.labelsize'] = 12
plt.rcParams['ytick.labelsize'] = 12

# Download the data and do some verifications

The data is located locally within this folder. 

In [4]:
%ls -l all/train.csv

-rwxr-xr-x 1 root root 3237288 Jul 19 03:51 all/train.csv*


In [5]:
import pandas as pd

In [6]:
full_train_data = pd.read_csv('all/train.csv')

In [7]:
train_data = pd.read_csv('all/train.csv')

In [17]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9557 entries, 0 to 9556
Columns: 143 entries, Id to Target
dtypes: float64(8), int64(130), object(5)
memory usage: 10.4+ MB


From Info it looks that Cabin have most number of missing values followed by Age. 

Above data shows that there are 177 passengers for whom the Age is unknown or null.
Also Cabin have lots of missing data around 687 rows(most)

In [16]:
pd.options.display.max_columns = 150
train_data.head()

Id      v2a1  hacdor  rooms  hacapo  v14a  refrig  v18q  v18q1  \
0  ID_279628684  190000.0       0      3       0     1       1     0    NaN   
1  ID_f29eb3ddd  135000.0       0      4       0     1       1     1    1.0   
2  ID_68de51c94       NaN       0      8       0     1       1     0    NaN   
3  ID_d671db89c  180000.0       0      5       0     1       1     1    1.0   
4  ID_d56d6f5f5  180000.0       0      5       0     1       1     1    1.0   

   r4h1  r4h2  r4h3  r4m1  r4m2  r4m3  r4t1  r4t2  r4t3  tamhog  tamviv  \
0     0     1     1     0     0     0     0     1     1       1       1   
1     0     1     1     0     0     0     0     1     1       1       1   
2     0     0     0     0     1     1     0     1     1       1       1   
3     0     2     2     1     1     2     1     3     4       4       4   
4     0     2     2     1     1     2     1     3     4       4       4   

   escolari  rez_esc  hhsize  paredblolad  paredzocalo  paredpreb  pareddes  \
0        10      NaN       1            1            0          0         0   
1        12      NaN       1            0            0          0         0   
2        11      NaN       1            0            0          0         0   
3         9      1.0       4            1            0          0         0   
4        11      NaN       4            1            0          0         0   

   paredmad  paredzinc  paredfibras  paredother  pisomoscer  pisocemento  \
0         0          0            0           0           1            0   
1         1          0            0           0           0            0   
2         1          0            0           0           1            0   
3         0          0            0           0           1            0   
4         0          0            0           0           1            0   

   pisoother  pisonatur  pisonotiene  pisomadera  techozinc  techoentrepiso  \
0          0          0            0           0          0               1   
1          0          0            0           1          1               0   
2          0          0            0           0          1               0   
3          0          0            0           0          1               0   
4          0          0            0           0          1               0   

   techocane  techootro  cielorazo  abastaguadentro  abastaguafuera  \
0          0          0          1                1               0   
1          0          0          1                1               0   
2          0          0          1                1               0   
3          0          0          1                1               0   
4          0          0          1                1               0   

   abastaguano  public  planpri  noelec  coopele  sanitario1  sanitario2  \
0            0       1        0       0        0           0           1   
1            0       1        0       0        0           0           1   
2            0       1        0       0        0           0           1   
3            0       1        0       0        0           0           1   
4            0       1        0       0        0           0           1   

   sanitario3  sanitario5  sanitario6  energcocinar1  energcocinar2  \
0           0           0           0              0              0   
1           0           0           0              0              1   
2           0           0           0              0              1   
3           0           0           0              0              1   
4           0           0           0              0              1   

   energcocinar3  energcocinar4  elimbasu1  elimbasu2  elimbasu3  elimbasu4  \
0              1              0          1          0          0          0   
1              0              0          1          0          0          0   
2              0              0          1          0          0          0   
3              0              0          1          0

In [11]:
train_data.corr()

v2a1    hacdor     rooms    hacapo      v14a    refrig  \
v2a1             1.000000 -0.091732  0.443461 -0.073509  0.033551  0.088970   
hacdor          -0.091732  1.000000 -0.233369  0.652594 -0.175011 -0.101965   
rooms            0.443461 -0.233369  1.000000 -0.213368  0.129183  0.130531   
hacapo          -0.073509  0.652594 -0.213368  1.000000 -0.150986 -0.124506   
v14a             0.033551 -0.175011  0.129183 -0.150986  1.000000  0.143143   
refrig           0.088970 -0.101965  0.130531 -0.124506  0.143143  1.000000   
v18q             0.278364 -0.084680  0.254256 -0.067529  0.036396  0.086002   
v18q1            0.302292 -0.049262  0.208919 -0.037414  0.011255 -0.070318   
r4h1            -0.081900  0.232508 -0.066578  0.226378 -0.054769 -0.047087   
r4h2            -0.002401  0.059313  0.267627  0.126645  0.018133 -0.022819   
r4h3            -0.059548  0.184857  0.195222  0.240056 -0.015552 -0.046860   
r4m1            -0.072791  0.268978 -0.032558  0.241452 -0.006370 -0.023502   
r4m2             0.048169  0.142458  0.241989  0.095545  0.038997  0.027832   
r4m3            -0.007806  0.264620  0.168503  0.212527  0.026510  0.008038   
r4t1            -0.103959  0.328901 -0.064789  0.306722 -0.039803 -0.046136   
r4t2             0.032631  0.134909  0.349206  0.152968  0.038295  0.001607   
r4t3            -0.043574  0.304282  0.245784  0.305857  0.007615 -0.025979   
tamhog          -0.064566  0.304884  0.240137  0.306289  0.007328 -0.026784   
tamviv          -0.051396  0.350948  0.254473  0.354389  0.010613 -0.022060   
escolari         0.288006 -0.122134  0.220284 -0.104846  0.036796  0.097733   
rez_esc         -0.086238  0.047466 -0.084857  0.081574 -0.022039 -0.116235   
hhsize          -0.064566  0.304884  0.240137  0.306289  0.007328 -0.026784   
paredblolad      0.290062 -0.136055  0.260981 -0.126550  0.058186  0.134937   
paredzocalo     -0.101152  0.074927 -0.057740  0.057683 -0.000611 -0.035913   
paredpreb       -0.174037 -0.027202 -0.151392 -0.036119  0.005202  0.022801   
pareddes        -0.037311  0.147449 -0.096855  0.164714 -0.134787 -0.087432   
paredmad        -0.134678  0.083620 -0.111619  0.083705 -0.059958 -0.162714   
paredzinc       -0.053664  0.155120 -0.070527  0.169976  0.008349 -0.053502   
paredfibras     -0.025562 -0.007622 -0.002567 -0.005961  0.002778  0.008057   
paredother      -0.015824 -0.007622 -0.021199 -0.005961  0.002778  0.008057   
...                   ...       ...       ...       ...       ...       ...   
bedrooms         0.297657 -0.209735  0.804070 -0.126310  0.102891  0.103223   
overcrowding    -0.233023  0.670727 -0.382364  0.530401 -0.141635 -0.113158   
tipovivi1             NaN -0.063173  0.170745 -0.026624  0.014706  0.005873   
tipovivi2        0.226177 -0.066533  0.093062 -0.052036  0.024248  0.058247   
tipovivi3       -0.226177  0.073560 -0.171118  0.035628 -0.007214  0.012889   
tipovivi4             NaN  0.151115 -0.103335  0.101824 -0.057659 -0.080603   
tipovivi5             NaN  0.010075 -0.114921  0.006048 -0.015246 -0.054244   
computer         0.272942 -0.067109  0.243198 -0.052486  0.010088  0.060654   
television       0.041303 -0.075837  0.226308 -0.064587  0.045742  0.080885   
mobilephone      0.065360 -0.059923  0.075540 -0.024046  0.053883  0.127180   
qmobilephone     0.108504  0.004046  0.404278  0.044267  0.049947  0.122118   
lugar1           0.231539 -0.042086  0.223771 -0.038896  0.027642  0.094955   
lugar2          -0.019377  0.017444 -0.073862  0.023858 -0.001824 -0.007976   
lugar3          -0.039030  0.025546 -0.130994 -0.005963  0.000709 -0.014484   
lugar4          -0.121150 -0.010172 -0.051941  0.000737 -0.020328 -0.006559   
lugar5          -0.128111  0.037182 -0.099481  0.023113 -0.021421 -0.064096   
lugar6          -0.143357  0.005289 -0.050428  0.024699 -0.005146 -0.075008   
area1            0.112759  0.027721  0.130286  0.008402 -0.007297  0.078661   
area2           -0.112759 -0.027721 -0.130286 -0.008402  0.007297 -0.07

In [ ]:
Number of Null Columns.

In [22]:
pd.options.display.max_rows = 150
train_data.isnull().sum()

Id                    0
v2a1               6860
hacdor                0
rooms                 0
hacapo                0
v14a                  0
refrig                0
v18q                  0
v18q1              7342
r4h1                  0
r4h2                  0
r4h3                  0
r4m1                  0
r4m2                  0
r4m3                  0
r4t1                  0
r4t2                  0
r4t3                  0
tamhog                0
tamviv                0
escolari              0
rez_esc            7928
hhsize                0
paredblolad           0
paredzocalo           0
paredpreb             0
pareddes              0
paredmad              0
paredzinc             0
paredfibras           0
paredother            0
pisomoscer            0
pisocemento           0
pisoother             0
pisonatur             0
pisonotiene           0
pisomadera            0
techozinc             0
techoentrepiso        0
techocane             0
techootro             0
cielorazo       

 ## Columns with Null values arranged descending. 

1. rez_esc -- 7928 
2. v18q1   -- 7342
3. v2a1    -- 6860
4. meaneduc -- 5
5. SQBmeaned -- 5
    
    